{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "7e101959",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "import itertools\n",
    "from itertools import product\n",
    "import re\n",
    "import cairo\n",
    "import math\n",
    "from IPython import display\n",
    "from IPython.display import SVG, display, Image\n",
    "import re\n",
    "# import argparse\n",
    "# #import zip\n",
    "\n",
    "# def get_args():\n",
    "#     parser = argparse.ArgumentParser(description=\"A program to input coverage limit\")\n",
    "#     parser.add_argument(\"-f\", help=\"fasta file\", required=True, type=str)\n",
    "#     parser.add_argument(\"-m\", help=\"motifs file\", required=True, type=str)\n",
    "#     return parser.parse_args()\n",
    "\n",
    "# args = get_args()\n",
    "fasta_filename = \"Figure_1.fasta\" # -f file carrying the sequences in fasta format\n",
    "motif_filename = \"Fig_1_motifs.txt\" # -m file carrying the motifs\n",
    "out_filename = fasta_filename.split('.')[0] + \".png\"\n",
    "\n",
    "# Dictionary which holds all the possible nucleotides represented by a specific IUPAC characters\n",
    "IPUAC_dict = {'A':['A'],'C':['C'], 'G':['G'],'T':['T'],'U':['U'],\n",
    "                'W':['A','T'],'S':['G','C'],'M':['A','C'],'K':['G','T'],\n",
    "                'R':['A','G'],'Y':['C','T'], 'B':['C','G','T'],\n",
    "                'D':['A','G','T'],'H':['A','C','T'],'V':['A','C','G'],\n",
    "                'N':['A','C','T','G']}\n",
    "\n",
    "def get_names(file):\n",
    "    '''\n",
    "    Takes a Fasta file and returns an array of just the sequence IDs\n",
    "    Input: fasta_file_handle\n",
    "    Output: list of sequence IDs\n",
    "    '''\n",
    "    file_handle = open(file, \"rt\")\n",
    "    names_list = []\n",
    "    lines = file_handle.readlines()\n",
    "    for line in lines:\n",
    "        if line.startswith(\">\"):\n",
    "            names_list.append(line.strip(\"\\n\")[1:])\n",
    "    return names_list\n",
    "\n",
    "def get_fasta_records(file):\n",
    "    '''\n",
    "    Takes a Fasta file where the sequences are split across multiple lines, and concatenates the sequence lines to a single string. Returns the sequence line.\n",
    "    Input: fasta_file_handle\n",
    "    Output: sequence\n",
    "    '''\n",
    "    file_handle = open(file, \"rt\")\n",
    "    record_list = []\n",
    "    for line in file_handle:\n",
    "        all_lines = file_handle.read()\n",
    "        record_array = all_lines.split('>')\n",
    "        record0_line_array = record_array[0].split('\\n') \n",
    "        record0_seq_line = ('').join(record0_line_array)\n",
    "        record_list.append(record0_seq_line)\n",
    "        for record in record_array[1:]:\n",
    "            record_line_array = record.split('\\n')  \n",
    "            seqs = record_line_array[1:]\n",
    "            seq_line = ('').join(seqs)\n",
    "            record_list.append(seq_line)\n",
    "    return record_list\n",
    "\n",
    "class Sequence:\n",
    "    def __init__(self, sequence):\n",
    "        self.sequence = sequence # to create the sequence object you pass in one of the fasta sequences\n",
    "        self.exon, self.seq_length = self.find_exon_introns(self.sequence)\n",
    "    \n",
    "    def find_exon_introns(self, sequence):\n",
    "        '''\n",
    "        This function takes as its input a DNA sequence with introns lowercase and exons UPPERCASE. It identifies the positions of the exon and the sequence length\n",
    "        Input: tttgtttccacagAAAAACCTCTTCAGGCACTGGTGCCGAGGACCCTAGgtatgactcacctgt\"\n",
    "        Output: [13, 48], 100\n",
    "        '''\n",
    "        seq_len = len(sequence)\n",
    "        exon = re.compile(\"[ACTGN]+\") # exons are represented by uppercase letters\n",
    "        exon_positions =  [[m.start(),m.end()] for m in exon.finditer(sequence)] # find the start and stop of the exon\n",
    "        exon_start_end = [exon_positions[0][0], (exon_positions[0][1]-1)]\n",
    "        return exon_start_end, seq_len    \n",
    "\n",
    "    \n",
    "    def get_motif_occurence_dict(self, motif_dict):\n",
    "        '''\n",
    "        This function takes as its input a fasta sequence and the dictionary of all possible motifs, and returns a dictionary of each motif and where it is located in the sequence.\n",
    "        Input: {\"YGCY\":[\"TGCT\",\"CGCT\",...]...}, and \"tttgtttccacagAAAAACCTCTTCAGGCACTGGTGCCGAGGACCCTAGgtatgactcacctgt\"\n",
    "        Output: {\"YGCY\":[[25,29], [31,35]...}\n",
    "        '''\n",
    "        motif_occurence_dict = {}\n",
    "        for motif in motif_dict.keys(): # for each motif in the dictionary created in the motif class (which has as keys all the motifs, and values all the possible motifs an ambiguous motif could represent)\n",
    "            motif_occurence_dict[motif] = [] # store the motif as a key in motif_occurence_dict, and set the value to an empty list\n",
    "            for possible_motif in motif_dict[motif]: # for each motif option in motif_dict (because motifs are ambiguous)\n",
    "                motif_pattern = re.compile(possible_motif) # create a regular expression pattern\n",
    "                motif_pattern_occurence = [[m.start(),m.end()] for m in motif_pattern.finditer((self.sequence).upper())] # search for occurrences of it in the gene sequence. NOTE: the motifs have been made uppercase, so sequence.upper is used\n",
    "                if len(motif_pattern_occurence) != 0: # if there is an occurence of the motif inside the gene\n",
    "                    motif_occurence_dict[motif].append(motif_pattern_occurence) # add it to the value in the dictionary\n",
    "        return motif_occurence_dict \n",
    "\n",
    "\n",
    "class Motifs:\n",
    "    def __init__(self, motif_file_name):\n",
    "        self.file = motif_file_name\n",
    "        self.motif_array = self.init_motif_array(motif_file_name)\n",
    "        self.motif_options_dict = self.find_all_motif_options()\n",
    "        \n",
    "    def __str__(self):\n",
    "        return str(self.motif_array)\n",
    "\n",
    "    def init_motif_array(self, file):\n",
    "        '''\n",
    "        Create an array from the motifs in the file \n",
    "        Input: motif_file_handle\n",
    "        Output: [motif1, motif2, motif3]\n",
    "        '''\n",
    "        motif_file = open(self.file, \"rt\")\n",
    "        all_motif_lines = motif_file.read()\n",
    "        motif_array = all_motif_lines.split('\\n')\n",
    "        motif_file.close()\n",
    "        return motif_array\n",
    "\n",
    "    def get_motif_array(self):\n",
    "        return self.motif_array\n",
    "        \n",
    "    def find_all_motif_options(self):\n",
    "        '''\n",
    "        Takes as input a single ambiguous motif and generates all possible motifs from it. Returns a dictionary with {motif: [[motif1.1, motif 1.2, motif 1.3],[motif2.1, motif 2.2, motif 2.3]}\n",
    "        Input: [motif1, motif2, motif3]\n",
    "        Output: {motif1: [[motif1.1, motif 1.2, motif 1.3], motif2: [motif2.1, motif 2.2, motif 2.3]}\n",
    "        '''\n",
    "        motif_dict = {}\n",
    "        for motif in self.motif_array:\n",
    "            motif = motif.upper() # make all letters uppercase\n",
    "            all_characters = []\n",
    "            for character in motif:\n",
    "                one_position_possible_chars = IPUAC_dict[character] # create an array for each position in the motif that holds every possible character that could be at that position according to its IUPAC\n",
    "                all_characters.append(one_position_possible_chars) \n",
    "            possible_strings_iters = list(product(*all_characters)) # itertools product function creates all possible combinations for each ambiguous motif, but it returns them as lists (C, C, T, G, ...)\n",
    "            possible_strings = []\n",
    "            for i in range(0, len((possible_strings_iters))):\n",
    "                possible_strings.append(''.join(possible_strings_iters[i])) # use join to turn the output of itertools product into strings  \n",
    "            motif_dict[motif] = possible_strings \n",
    "        return motif_dict\n",
    "\n",
    "    def get_all_motif_options(self):\n",
    "        return self.motif_options_dict \n",
    "\n",
    "class Plot:\n",
    "    def __init__(self, plot_width, plot_height, file_name):\n",
    "        self.my_motif_colors = {} # dictionary which holds motifs and their rgb values\n",
    "        self.width = plot_width \n",
    "        self.height = plot_height \n",
    "        self.file = file_name # file name for the png file\n",
    "        self.surface = cairo.SVGSurface(self.file, self.width, self.height) # create png with w/h dimensions\n",
    "        self.context = cairo.Context(self.surface) # create context on which to draw\n",
    "        \n",
    "    def set_motif_colors(self, motif_string, colors): \n",
    "        '''\n",
    "        This function takes as its input a motif string and a set of rgb colors, and creates a dictionary with the motif string as key and rgb numbers as values\n",
    "        Input: \"GCUAG\", [1,0.8,0.2 ]\n",
    "        Output: {\"GCUAG\": [1,0.8,0.2 ], ...}\n",
    "        '''\n",
    "        self.my_motif_colors[motif_string] = colors\n",
    "\n",
    "    def get_colors_for_motif(self, motif_string): \n",
    "        '''\n",
    "        This function takes as input a motif string and gets the rgb values for that motif string from the my_motif_colors dictionary and store them as color_r, color_b, color_g\n",
    "        Input: a motif\n",
    "        Outpu: 0.1, 0.6, 0.8 (some RGB colors)\n",
    "        '''\n",
    "        color_r = self.my_motif_colors[motif_string][0]\n",
    "        color_g = self.my_motif_colors[motif_string][1]\n",
    "        color_b = self.my_motif_colors[motif_string][2]\n",
    "        return color_r, color_g, color_b\n",
    "    \n",
    "    def create_background_color(self): # set the backgroun color to white\n",
    "        '''\n",
    "        This function creates a white rectangle that is the height and width of the surface because cairo's default context background is black/transparent\n",
    "        '''\n",
    "        self.context.rectangle(0,0,self.width,self.height)\n",
    "        self.context.set_source_rgb(1, 1, 1)\n",
    "        self.context.fill()\n",
    "    \n",
    "    def plot_motifs(self, motif_occurence_dict, sequence_obj, seq, plot_index):\n",
    "        '''\n",
    "        This function creates the motif plots \n",
    "        Input: motif_occurence_dict (a dictionary of each motif and where it occurs in a sequence), a sequence object, a gene sequence, the index of the gen sequence in the gene sequence array)\n",
    "        Output: A plot which creates color-coded blocks of the motifs on a sequence, which is represented as a black line\n",
    "        '''\n",
    "        exon = sequence_obj.find_exon_introns(seq)[0] # get the exon start and stop from the sequence object\n",
    "        exon_start = exon[0] # the start is the first item in the exon list\n",
    "        exon_len = (exon[1]-exon[0]) # the length of the exon is the second item - first item in the exon list\n",
    "        seq_len = sequence_obj.find_exon_introns(seq)[1] # get the sequence length from the sequence object\n",
    "        self.context.set_source_rgb(0,0,0) # set the color of the text labels to be black\n",
    "        self.context.move_to(50,(100 + (200*(plot_index-1)))) # start the text at 50,100, and increment this start by 200 for every subsequent sequence\n",
    "        self.context.set_font_size(20) \n",
    "        self.context.show_text(names_list[plot_index-1])\n",
    "        self.context.set_line_width(3) # width of the line representing the sequence\n",
    "        self.context.set_source_rgb(0,0,0) # color of sequence line = black\n",
    "        self.context.move_to(100, (200*plot_index)) # origin of the line, left-most point\n",
    "        self.context.line_to((100+seq_len),(200*plot_index)) # right-most point is sequence length - start position\n",
    "        self.context.stroke() \n",
    "        self.context.rectangle((100+exon_start),(150 + (200*(plot_index-1))),exon_len,(100)) # Create the exon box (x, y, width, height)\n",
    "        self.context.set_source_rgba(0,0,0, 0.1) # make it transparent\n",
    "        self.context.fill()\n",
    "        \n",
    "        \n",
    "\n",
    "        for motif in motif_occurence_dict.keys():\n",
    "        # for each motif : YCGY, GCAUG, CATAG, YYYYYYYYYY\n",
    "            motif_blocks = [] # position array like this: [[79, 89], [80, 90], [76, 86], [78, 88], [75, 85], [77, 87], [74, 84], [73, 83]]\n",
    "            for location_array in motif_occurence_dict[motif]: # for the locations of each motif\n",
    "                for location in location_array: \n",
    "                    start = location[0] # get the start of the motif\n",
    "                    length = (location[1]-location[0]) # get the length of the motif\n",
    "                    motif_blocks.append(location)\n",
    "                    color_r, color_g, color_b = self.get_colors_for_motif(motif) # get the color assigned to that motif \n",
    "                    self.context.set_source_rgba(color_r, color_g, color_b, 0.5) \n",
    "                    self.context.rectangle((100+start),(150 + (200*(plot_index-1))),length,(100)) # create a rectangle (x, y, width, height) of the right size and color\n",
    "                    self.context.fill()\n",
    "                    self.context.stroke()\n",
    "            \n",
    "        \n",
    "    def create_key(self, motif, n):\n",
    "        self.context.rectangle(850, (100 + 40*n), 20, 20) # (x, y, width, height)\n",
    "        color_r, color_g, color_b = self.get_colors_for_motif(motif)\n",
    "        self.context.set_source_rgba(color_r, color_g, color_b, 0.5) # blue\n",
    "        self.context.fill()\n",
    "        \n",
    "        self.context.move_to(880,(120 + 40*n))\n",
    "        self.context.set_font_size(20)\n",
    "        self.context.set_source_rgb(0,0,0) # black\n",
    "        self.context.show_text(motif)\n",
    "\n",
    "        self.surface.write_to_png(out_filename)\n",
    "\n",
    "\n",
    "motifs_obj = Motifs(motif_filename) # create motif object\n",
    "motif_list = motifs_obj.get_motif_array() # [\"YGCY\",\"GCAUG\",\"CATAG\",\"YYYYYYYYYY\"]\n",
    "motif_dict = motifs_obj.get_all_motif_options()\n",
    "\n",
    "record_list = get_fasta_records(fasta_filename)\n",
    "names_list = get_names(fasta_filename)\n",
    "\n",
    "myplot = Plot(1200, 1200, out_filename)\n",
    "myplot.create_background_color()\n",
    "\n",
    "color_list = [[1,0.6,0.3],[1,0.8,0.2 ],[0,0.2,0.8 ], [0.2,0.9,0.8],[0.5,0.3,1]] # yellow, blue, green, orange, # purple\n",
    "\n",
    "for i in zip(motif_list, color_list):\n",
    "    myplot.set_motif_colors(i[0].upper(), i[1]) \n",
    "    \n",
    "    \n",
    "n = 0\n",
    "for i in record_list:\n",
    "    n += 1\n",
    "    sequence = Sequence(i) # create sequence object\n",
    "    MOdict = sequence.get_motif_occurence_dict(motif_dict)\n",
    "    myplot.plot_motifs(MOdict, sequence, i, n)\n",
    "    \n",
    "for i in range(len(motif_list)):\n",
    "    n = i\n",
    "    motif = motif_list[i].upper()\n",
    "    myplot.create_key(motif, n)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
